src/Workspaces/Core/Portable/ObsoleteSymbol/AbstractObsoleteSymbolService.cs

```diff
@@ -0,0 +1,176 @@
+
+internal abstract class AbstractObsoleteSymbolService(int? dimKeywordKind) : IObsoleteSymbolService
+{
+    private readonly int? _dimKeywordKind = dimKeywordKind;
+
+    protected virtual void ProcessDimKeyword(ArrayBuilder<TextSpan> result, SemanticModel semanticModel, SyntaxToken token, CancellationToken cancellationToken)
+    {
+        // Take no action by default
+    }
+
+    public async Task<ImmutableArray<TextSpan>> GetLocationsAsync(Document document, ImmutableArray<TextSpan> textSpans, CancellationToken cancellationToken)
+    {
+        var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+        var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        using var _2 = ArrayBuilder<TextSpan>.GetInstance(out var result);
+        var semanticModel = compilation.GetSemanticModel(root.SyntaxTree);
+
+        foreach (var span in textSpans)
+        {
+            Recurse(span, semanticModel);
+        }
+
+        result.RemoveDuplicates();
+        return result.ToImmutableAndClear();
+
+        void Recurse(TextSpan span, SemanticModel semanticModel)
+        {
+            using var _ = ArrayBuilder<SyntaxNode>.GetInstance(out var stack);
+
+            // Walk through all the nodes in the provided span.  Directly analyze local or parameter declaration.  And
+            // also analyze any identifiers which might be reference to locals or parameters.  Note that we might hit
+            // locals/parameters without any references in the span, or references that don't have the declarations in 
+            // the span
+            stack.Add(root.FindNode(span));
+
+            // Use a stack so we don't blow out the stack with recursion.
+            while (stack.Count > 0)
+            {
+                var current = stack.Last();
+                stack.RemoveLast();
+
+                if (current.Span.IntersectsWith(span))
+                {
+                    var tokenFromNode = ProcessNode(semanticModel, current);
+
+                    foreach (var child in current.ChildNodesAndTokens())
+                    {
+                        if (child.IsNode)
+                        {
+                            stack.Add(child.AsNode()!);
+                        }
+
+                        var token = child.AsToken();
+                        if (token != tokenFromNode)
+                            ProcessToken(semanticModel, child.AsToken());
+
+                        foreach (var trivia in token.LeadingTrivia)
+                        {
+                            if (trivia.HasStructure)
+                            {
+                                stack.Add(trivia.GetStructure()!);
+                            }
+                        }
+
+                        foreach (var trivia in token.TrailingTrivia)
```

Comment:
nit.  extract helper fort ehse loops.  so you can just call it both with Leading/TrailingTrivia

Reply:
➡️ Extracted helper